crypto/asn1/a_time.c

```diff
@@ -200,16 +393,8 @@ int ASN1_TIME_to_tm(const ASN1_TIME *s, struct tm *tm)
         return 0;
     }
 
-    if (s->type == V_ASN1_UTCTIME) {
-        memset(tm, 0, sizeof(*tm));
-        return asn1_utctime_to_tm(tm, s);
-    }
-    if (s->type == V_ASN1_GENERALIZEDTIME) {
-        memset(tm, 0, sizeof(*tm));
-        return asn1_generalizedtime_to_tm(tm, s);
-    }
-
-    return 0;
+    memset(tm, 0, sizeof(*tm));
```

Comment:
So the problem is how should the `ASN1_TIME_to_tm` be designed. There might be 2 choices:

1. set the listed (in doc) fields of `tm` to proper values, and zero all others.
2. set the listed fields of `tm` to proper values, and leave other fields as-is.

For both choices, it's only applied if the 'TIME' `s` is parsed successfully, otherwise leave `tm` untouched.

I prefer the first choice, and I guess this is what `ASN1_TIME_to_tm` intends to be. If a user wants to just fill the listed fields, he/she could call something like `ASN1_TIME_to_tm(s, &tmp)` and copy the fields of `tmp` to other `tm`/places by his/her own. That's not the scope covered by this API.

So the `memset` stuff in `ASN1_TIME_to_tm` should be removed and documentation of `ASN1_TIME_to_tm` should be modified to match this behavior either.

And also, in this scenario, if a NULL `tm` is passed to `ASN1_TIME_to_tm` and finally is passed to `asn1_time_to_tm`, the `ASN1_TIME_to_tm` will perform in a 'format-checking-only' mode, which is similar to what `ASN1_TIME_set_string` does. This feature should also be documented, of course.

Makes sense to what you suggested?